PF-431, PF-592, PF-605 - Resource enumeration

ddietterich opened the pull request:
- cleanup OpenAPI for consistent naming and contents for resource returns and enumerated resource list
 - remove stray uses of deprecated schemas
 - add enumeration endpoint
 - restructure/rename per-object converters from Wsm classes into Api classes
 - teach controlled bucket to return the whole resource on create and get
 - added enumerateResources testrunner test
 - added more logging and enabled retries for delete workspace authz step
 - there is a problem...

Something isn't working right with the permissions on controlled resources
Either the query in ResourceDao is broken or the permissions are not getting
setup properly. The test has a workaround to only create and enumerate
referenced resources.

Since the client is changed, the integration tests do not pass in the GHA. The FullIntegration suite (with the workaround) passes on my personal environment.

jaycarlton commented on src/main/java/bio/terra/workspace/service/resource/WsmResource.java:
```diff
@@ -97,6 +98,24 @@ public CloningInstructions getCloningInstructions() {
 
+  /**
+   * Create the part of the common resource description that comes from this class. Subclasses
+   * should override this and fill in their specific info.
```
Indicate in the comment that the info to be specified is on a field in the `ApiResourceMetadata` class, and that they shouldn't extend that class to hold their state.

wchamber replied:
Should we have a protected abstract method for subclasses to override? Would make it harder to forget.

jaycarlton commented on src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResource.java:
```diff
@@ -63,6 +66,20 @@ public ManagedByType getManagedBy() {
+            .privateResourceUser(
+                // TODO: PF-616 figure out how to supply the assigned user's role
+                new ApiPrivateResourceUser().userName(assignedUser));
+    metadata.controlledResourceMetadata(controlled);
```
Do we stop here, or can we keep going with the subclass's metadata?

jaycarlton reviewed (commented):
quick pass

ddietterich commented on src/main/resources/api/service_openapi.yaml:
```diff
@@ -1,3 +1,27 @@
+# Each cloud resource has a structure named '{resource-type}Resource';
+# for example, GcsBucketResource. It has two properties: 'metadata' and 'attributes'
+#
+# The 'metadata' property is a reference to the common metadata structure for all resource types:
```
It is not right now.

ddietterich commented on src/main/resources/api/service_openapi.yaml:
```diff
@@ -1,3 +1,27 @@
+# data references interface that we abandoned in favor of the unified resources interface.
+#
+# CLOUD RESOURCE STRUCTURE
+# Each cloud resource has a structure named '{resource-type}Resource';
```
Thanks for the reminder. I updated the documentation and did the rename to add the cloud platform

wchamber commented on src/main/java/bio/terra/workspace/db/ResourceDao.java:
```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
+    }
+
+    // There are three cases for the stewardship type filter
+    // 1. If it is REFERENCED, then we ignore idList and just filter
+    //    for referenced resources.
+    // 2. If it is CONTROLLED, and the idList is not empty, then we filter for
+    //    CONTROLLED and require that the resources be in the idList.
+    // 3. If it is null, then we want the OR of 1 and 2
```
This seems to me like we want different public enumerate methods on ResourceDao. When there are some arguments that are only valid with certain values of arguments, I think it's best to define a Java API where invalid argument combinations are less allowed. While our HTTP API can handle all the methods, and do the appropriate request validation on arguments, I don't think we should restrict our Java methods to be 1 to 1 because of that.
Maybe a enumerateControlledResources methods when we want to specify controlledResourceIds.

At least let's explain in the javadoc here and on ResourceService how these parameters interact together.

ddietterich replied:
I expanded the javadoc comment to try to clarify. The complexity here is due to a problem I found when passing in the id list. If the list is empty, the resulting SQL is invalid. So instead of simply adding one more filter, like resource type, I end up with this more complex query construction.

I have a bias to invest in the query generation and keep the database work to a single scan/single transaction.

wchamber commented on src/main/java/bio/terra/workspace/service/iam/SamService.java:
```diff
@@ -398,6 +398,25 @@ public void deleteControlledResource(
+  public List<UUID> listControlledResourceIds(
+      AuthenticatedUserRequest userReq, String controlledResourceName) {
+    ResourcesApi resourceApi = samResourcesApi(userReq.getRequiredToken());
+    List<UUID> workspaceIds = new ArrayList<>();
```
controlledResourceIds?

wchamber commented on src/main/java/bio/terra/workspace/service/iam/SamService.java:
```diff
@@ -398,6 +398,25 @@ public void deleteControlledResource(
+   * across the four types of controlled resources.
+   */
+  @Traced
+  public List<UUID> listControlledResourceIds(
+      AuthenticatedUserRequest userReq, String controlledResourceName) {
```
I'm a little confused by the primitive String controlledResourceName and UUID return value. What kind of "name" are we storing in Sam? What kind of UUID, and how does it relate to the ResourceDao, does this return?
Bigger scope than this PR, but I think it would be very useful to strongly type our UUIDs and String "types" of data. For UUID, I know we have at least a "WorkspaceId" and a "ResourceId".  The Sam service also has "resource" as a part of its terminology and API, see SamService#isAuthorized: It's unclear to me what a `String resourceId` represent.

wchamber commented on src/main/java/bio/terra/workspace/service/resource/WsmResourceType.java:
```diff
@@ -50,6 +68,25 @@ public static WsmResourceType fromSql(String dbString) {
+   * @param apiResourceType incoming resource type or null
+   * @return valid resource type; null if input is null
+   */
+  public static WsmResourceType fromApi(ApiResourceType apiResourceType) {
```
nit: add `@Nullable` to the return type.

wchamber commented on src/main/java/bio/terra/workspace/service/resource/WsmResourceType.java:
```diff
@@ -50,6 +68,25 @@ public static WsmResourceType fromSql(String dbString) {
   }
 
+  /**
+   * Convert from api type to WsmResourceType. In some contexts, the API input can be null, so if
+   * the input is null we return null and leave it to the caller to raise any error.
```
"the null returns null, but invalid throws" is a little subtle to me. Do we really need 2 forms of "could not convert?" Can the clients do their own null checking?

wchamber commented on src/main/java/bio/terra/workspace/service/resource/controlled/AccessScopeType.java:
```diff
@@ -2,27 +2,24 @@
-      ApiControlledResourceCommonFields.AccessScopeEnum apiAccessScope) {
+  public static AccessScopeType fromApi(ApiAccessScope apiAccessScope) {
     if (apiAccessScope == null) {
       throw new MissingRequiredFieldException("Valid accessScope is required");
```
This enum conversion throws on null, but ApiResourceType doesn't? Consistency seems good.

wchamber commented on src/main/java/bio/terra/workspace/service/workspace/flight/DeleteWorkspaceAuthzStep.java:
```diff
@@ -31,8 +34,14 @@ public StepResult doStep(FlightContext flightContext) throws RetryException {
     } catch (SamApiException e) {
       // Stairway steps may run multiple times, so this may already have been deleted.
       // For all other errors we should always retry because there's no way to undo a Sam delete.
+      // TODO: there is no retry rule defined for this step, so it is never retried.
```
It looks like you've now added one to WorkspaceDeleteFlight?

wchamber reviewed (commented):
It would be nice to have broken this into smaller PRs, e.g. rename all the models, rework where the toApi functions are (possibly in multiple PRs), the bucket changes, add the new enumerate endpoint + tests for it.
Here's some partial comments. I don't have any objections to the thrust of the PR, but it's possible I missed something.

ddietterich commented on src/main/resources/api/service_openapi.yaml:
```diff
@@ -1,3 +1,27 @@
+# The 'metadata' property is a reference to the common metadata structure for all resource types:
+#  $ref: '#/components/schemas/ResourceMetadata'
+#
+# The 'attributes' property is a reference to a unique structure named '{resource-type}Attributes';
```
I don't think we want to assume that. Did you have something specific in mind?

jaycarlton replied:
I don't know where I got that idea.

jaycarlton commented on src/main/java/bio/terra/workspace/db/ResourceDao.java:
```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
+    // is defined in exactly one place.
+    boolean includeReferenced = (stewardshipType == null || stewardshipType == REFERENCED);
+    boolean includeControlled =
+        (!idList.isEmpty() && (stewardshipType == null || (stewardshipType == CONTROLLED)));
```
nit: you don't need the second open paren before `stewardshipType`

jaycarlton replied:
then why not also do it before the `== null` case?

jaycarlton commented on src/main/java/bio/terra/workspace/db/ResourceDao.java:
```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
+        "(stewardship_type = :controlled_resource AND resource_id IN (:id_list))";
+
+    sb.append(" AND ");
+    if (includeReferenced && includeControlled) {
```
If we get any fancier, we should look into dedicated query building classes/DSLs.

jaycarlton commented on src/main/java/bio/terra/workspace/db/ResourceDao.java:
```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
+      params.addValue("id_list", idList);
+    } else {
+      // Nothing is included, so we return an empty result
+      return new ArrayList<WsmResource>();
```
`Collections.emptyList()` is cleaner I think

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/testscripts/EnumerateResources.java:
```diff
@@ -0,0 +1,289 @@
+                        && m.getResourceType() == ResourceType.GCS_BUCKET))
+            .count();
+    logger.info("Counted {} controlled buckets created", expectedControlledBuckets);
+    long actualControlledBucket = controlledBucketList.getResources().size();
```
*actualControlledBuckets

zloery commented on src/main/java/bio/terra/workspace/service/iam/SamService.java:
```diff
@@ -398,6 +398,25 @@ public void deleteControlledResource(
+    ResourcesApi resourceApi = samResourcesApi(userReq.getRequiredToken());
+    List<UUID> workspaceIds = new ArrayList<>();
+    try {
+      for (var resourceAndPolicy : resourceApi.listResourcesAndPolicies(controlledResourceName)) {
```
this needs to call `listResourcesAndPoliciesV2`. Otherwise, the response won't include the resources a user has inherited permission on, which is most controlled resource

zloery reviewed (commented):
first pass of comments

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/utils/ResourceMaker.java:
```diff
@@ -0,0 +1,172 @@
+            .metadata(
+                new ReferenceResourceCommonFields()
+                    .cloningInstructions(CloningInstructionsEnum.NOTHING)
+                    .description("Descriptiono of " + name)
```
*description

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/utils/ResourceMaker.java:
```diff
@@ -0,0 +1,172 @@
+      createdBucket = resourceApi.getCreateBucketResult(workspaceId, jobId);
+    }
+    logger.info("Create bucket status is {}", createdBucket.getJobReport().getStatus().toString());
+    if (createdBucket.getJobReport().getStatus() != JobReport.StatusEnum.SUCCEEDED) {
```
can we use `assert()` in this kind of test?

jaycarlton replied:
Some lightweight assertion library that would throw an assertion failure in the test seems helpful. I suspect others will have opinions here.

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/utils/ResourceMaker.java:
```diff
@@ -0,0 +1,172 @@
+    }
+    logger.info("Delete bucket status is {}", result.getJobReport().getStatus().toString());
+    if (result.getJobReport().getStatus() != JobReport.StatusEnum.SUCCEEDED) {
+      throw new RuntimeException("Delte bucket failed: " + result.getErrorReport().getMessage());
```
*Delete

ddietterich commented on src/main/java/bio/terra/workspace/app/controller/ResourceController.java:
```diff
@@ -0,0 +1,148 @@
+              break;
+            }
+
+          case BIG_QUERY_DATASET:
```
I added documentation

jaycarlton commented on src/main/java/bio/terra/workspace/db/ResourceDao.java:
```diff
@@ -134,6 +141,80 @@ public boolean deleteResource(UUID workspaceId, UUID resourceId) {
+    // 3. If it is null, then we want the OR of 1 and 2
+    // Note that we supply the constant value as a parameter so that the SQL form
+    // is defined in exactly one place.
+    boolean includeReferenced = (stewardshipType == null || stewardshipType == REFERENCED);
```
Oh I completely misread it. Never mind.

jaycarlton commented on src/main/java/bio/terra/workspace/service/resource/controlled/AccessScopeType.java:
```diff
@@ -2,27 +2,24 @@
 public enum AccessScopeType {
-  ACCESS_SCOPE_SHARED("ACCESS_SHARED"),
-  ACCESS_SCOPE_PRIVATE("ACCESS_PRIVATE");
+  ACCESS_SCOPE_SHARED("ACCESS_SHARED", ApiAccessScope.SHARED_ACCESS),
```
Oof.

jaycarlton commented on src/main/java/bio/terra/workspace/service/resource/controlled/AccessScopeType.java:
```diff
@@ -46,4 +43,12 @@ public static AccessScopeType fromSql(String dbString) {
+    return dbString;
+  }
+
+  public ApiAccessScope toApiModel() {
```
That's reasonable. FWIW I like this model; it was more of a devil's advocate question.

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/testscripts/CreateGetDeleteControlledGcsBucket.java:
```diff
@@ -171,36 +144,35 @@ public void doUserJourney(TestUserSpecification testUser, WorkspaceApi workspace
   }
 
-  private CreatedControlledGcsBucket createBucketAttempt(ControlledGcpResourceApi resourceApi)
+  private CreatedControlledGcpGcsBucket createBucketAttempt(ControlledGcpResourceApi resourceApi)
```
Maybe `createBucketOrThrow` or something more descriptive?

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/testscripts/EnumerateResources.java:
```diff
@@ -0,0 +1,289 @@
+                ResourceMaker.makeBigQueryReference(referencedGcpResourceApi, workspaceId, name);
+            resourceList.add(resource.getMetadata());
+            /*
+            TODO: REVIEWERS: something isn't working right with the permissions on controlled resources
```
That's why we keep you around.

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/utils/CloudContextMaker.java:
```diff
@@ -0,0 +1,47 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CloudContextMaker {
```
this may not technically be a factory in the sense of Factory Pattern, so maker might be more accurate. https://www.tutorialspoint.com/design_pattern/factory_pattern.htm

ddietterich commented:
I think I have addressed most all of the review comments or made tickets for things to do later.
I did add a unit test for makeApiDescription as suggested by Jay.

The testrunner integration tests will not pass in GHA, because I have changed the API and the client doesn't match.
I ran them myself and all is well.

Please take a look. There are two meaningful commits to look at. The first is the cleanup from the code review. The second is getting the new enumerate integration test to work and adding the new unit test.

zloery commented on src/main/java/bio/terra/workspace/service/iam/SamService.java:
```diff
@@ -403,18 +404,20 @@ public void deleteControlledResource(
    * across the four types of controlled resources.
```
nit: this method doesn't build a list across resource types. it returns the list for a single type, and the aggregation happens elsewhere

zloery reviewed (approved):
LGTM

jaycarlton commented on src/test/java/bio/terra/workspace/service/resource/MakeApiResourceDescriptionTest.java:
```diff
@@ -0,0 +1,175 @@
+                managedByType,
+                bucketName);
+
+            ApiResourceDescription resourceDescription = resourceController.makeApiResourceDescription((WsmResource) resource);
```
why is this cast necessary?

ddietterich replied:
It isn't. Even IntelliJ knows :)

jaycarlton commented on workspace-manager-clienttests/src/main/java/scripts/testscripts/EnumerateResources.java:
```diff
@@ -0,0 +1,279 @@
+    for (ResourceDescription desc : descList) {
+      ResourceMetadata metadata = desc.getMetadata();
+
+      String access = "<null>";
```
nit: define this string
